File: Algorithm/Python/082. Remove Duplicates from Sorted List II.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-05-20 09:44:41
# @Last modified by:   WuLC
# @Last Modified time: 2016-05-20 09:50:45
# @Email: liangchaowu5@gmail.com

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# two pointers,point to closest different values
class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None or head.next == None:
            return head
        dummy = ListNode(0)
        p1, p2= head, head.next
        while p2 :
            if p2.val != p1.val:
                if p1.next == p2:
                    if dummy.next == None:
                        curr = p1
                        dummy.next = curr
                    else:
                        curr.next = p1
                        curr = curr.next
                p1 = p2
            p2 = p2.next
            
        if p1.next == None:        # the last element is valid
            if dummy.next == None: # only the last element is valid in the list 
                dummy.next = p1
            else:
                curr.next = p1
        else:                   
            if dummy.next != None:
                curr.next = None
        return dummy.next       
        

File: Algorithm/Java/082. Remove Duplicates from Sorted List II.java
/*
* @Author: WuLC
* @Date:   2017-05-26 20:13:59
* @Last Modified by:   WuLC
* @Last Modified time: 2017-05-26 20:14:37
* @Email: liangchaowu5@gmail.com
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */


// store duplicate numbers in a set
// O(n) time, O(n) space
public class Solution 
{
    public ListNode deleteDuplicates(ListNode head) 
    { 
        if  (head == null || head.next == null) return head;
    
        // find duplicate elements
        ListNode curr = head;
        Set<Integer> duplicate = new HashSet<Integer>();
        while ( curr.next != null)
        {
            if (curr.val == curr.next.val) duplicate.add(curr.val);
            curr = curr.next;
        }
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode p1 = dummy;
        curr = head;
        while (curr != null)
        {
            if (!duplicate.contains(curr.val))
            {
                p1.next = curr;
                p1 = p1.next;
            }
            curr = curr.next;
        }
        p1.next = null;
        return dummy.next;
    }
}